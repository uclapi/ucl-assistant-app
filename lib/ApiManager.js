import axios from "axios";
import moment from "moment";
import { encode as btoa } from "base-64";
import {
  ROOMS_URL,
  WORKSPACES_URL,
  EQUIPMENT_URL,
  ROOMBOOKINGS_URL,
} from "../constants/API";

const roomsController = {
  search: async (token = null, query) => {
    if (query && query.length < 3) {
      return {};
    }
    try {
      const results = await axios.get(ROOMS_URL, {
        params: {
          query,
        },
        headers: {
          authorization: `Bearer ${token}`,
          "uclapi-roombookings-version": "1",
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.rooms;
    } catch (error) {
      throw error;
    }
  },
  getEquipment: async (token = null, { roomid, siteid }) => {
    if (!roomid || !siteid) {
      throw new Error("Must specify roomid and siteid");
    }
    try {
      const results = await axios.get(EQUIPMENT_URL, {
        params: {
          roomid,
          siteid,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.equipment;
    } catch (error) {
      throw error;
    }
  },
  getBookings: async (
    token = null,
    { roomid, siteid, date = moment().format("YYYYMMDD") },
  ) => {
    if (!roomid || !siteid) {
      throw new Error("Must specify roomid and siteid");
    }
    try {
      const results = await axios.get(ROOMBOOKINGS_URL, {
        params: {
          roomid,
          siteid,
          date,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.bookings;
    } catch (error) {
      throw error;
    }
  },
};

const workspacesApi = axios.create({
  baseURL: WORKSPACES_URL,
});

const workspacesController = {
  getLiveImage: async (token = null, { surveyId, mapId }) => {
    try {
      const result = await workspacesApi.get("/getliveimage/map.svg", {
        params: {
          survey_id: surveyId,
          map_id: mapId,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
        responseType: "arraybuffer",
      });
      const image = btoa(
        new Uint8Array(result.data).reduce(
          (data, byte) => data + String.fromCharCode(byte),
          "",
        ),
      );
      // there may be no need to base64 SVGs
      // but I have yet to test this extensively
      // https://css-tricks.com/probably-dont-base64-svg/
      return `data:image/svg+xml;base64,${image}`;
    } catch (error) {
      throw error;
    }
  },
};

export default {
  rooms: roomsController,
  workspaces: workspacesController,
};
