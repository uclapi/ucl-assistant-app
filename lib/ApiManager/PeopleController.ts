import axios from 'axios'
import { PEOPLE_URL } from '../../constants/API'
import ErrorManager from '../ErrorManager'

const peopleApi = axios.create({
  baseURL: PEOPLE_URL,
})

class PeopleController {
  static search = async (token, query) => {
    if (query && query.length <= 3) {
      return null
    }
    try {
      const { data } = await peopleApi.get(
        `/?query=${encodeURIComponent(query)}`,
        {
          headers: {
            authorization: `Bearer ${token}`,
          },
        },
      )
      return data.content.people
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  }

  static fetchPerson = async (token, email) => {
    try {
      const results = await PeopleController.search(token, email)
      if (results.length > 0) {
        return results[0]
      }
      throw new Error(`No such person found`)
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  }
}

export default PeopleController
