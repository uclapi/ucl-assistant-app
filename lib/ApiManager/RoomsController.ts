import axios from "axios"
import {
  EQUIPMENT_URL, FREE_ROOMS_URL, ROOMBOOKINGS_URL, ROOMS_URL, SITES_URL,
} from "../../constants/API"
import ErrorManager from "../ErrorManager"
import LocalisationManager from "../LocalisationManager"

class roomsController {
  static getBookings = async (
    token = null,
    {
      roomid,
      siteid,
      date = LocalisationManager.getMoment().format(`YYYYMMDD`),
    },
  ) => {
    if (!roomid || !siteid) {
      throw new Error(`Must specify roomid and siteid`)
    }
    try {
      const results = await axios.get(ROOMBOOKINGS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          date,
          roomid,
          siteid,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.bookings
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  }

  static getEmptyRooms = async (
    token = null,
    {
      startDateTime = (new Date()).toISOString(),
      endDateTime = (
        LocalisationManager
          .getMoment()
          .add(1, `hours`)
          .toISOString()
      ),
    } = {},
  ) => {
    try {
      const results = await axios.get(FREE_ROOMS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          endDateTime,
          startDateTime,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.free_rooms
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  }

  static getEquipment = async (token = null, { roomid, siteid }) => {
    if (!roomid || !siteid) {
      throw new Error(`Must specify roomid and siteid`)
    }
    try {
      const results = await axios.get(EQUIPMENT_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          roomid,
          siteid,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.equipment
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  }

  static getSites = async (token = null) => {
    try {
      const results = await axios.get(SITES_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.sites
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  }

  static search = async (token = null, query = ``) => {
    if (query && query.length < 3) {
      return {}
    }
    try {
      const results = await axios.get(ROOMS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          query,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.rooms
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  }
}

export default roomsController
