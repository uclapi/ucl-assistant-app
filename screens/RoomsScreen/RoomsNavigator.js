import { createStackNavigator } from 'react-navigation-stack'

import EmptyRoomsScreen from './EmptyRoomsScreen'
import RoomsFavouritesScreen from './RoomsFavouritesScreen'
import RoomsSearchScreen from './RoomsSearchScreen'

const RoomsNavigator = createStackNavigator(
  {
    EmptyRooms: {
      screen: EmptyRoomsScreen,
    },
    RoomsFavourites: {
      screen: RoomsFavouritesScreen,
    },
    RoomsSearch: {
      screen: RoomsSearchScreen,
    },
  },
  {
    defaultNavigationOptions: () => ({
      headerTitleStyle: {
        fontFamily: `apercu`,
        fontWeight: `normal`,
      },
    }),
    initialRouteName: `RoomsFavourites`,
  }
)

export default RoomsNavigator
