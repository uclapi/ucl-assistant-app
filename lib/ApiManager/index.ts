import Constants from 'expo-constants'
import PeopleController from './PeopleController'
import RoomsController from './RoomsController'
import TimetableController from './TimetableController'
import WorkspacesController from './WorkspacesController'

const clientId = Constants.manifest.extra?.uclapi?.clientId

export default {
  people: PeopleController,
  rooms: RoomsController,
  timetable: TimetableController,
  workspaces: WorkspacesController,
  clientId,
}
