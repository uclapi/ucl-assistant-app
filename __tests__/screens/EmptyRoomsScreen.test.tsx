/**
 * @jest-environment jsdom
 */
import "react-native"

import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'

import ApiManager from "../../lib/ApiManager"
import { EmptyRoomsScreen } from '../../screens/RoomsScreen/EmptyRoomsScreen'

Enzyme.configure({ adapter: new Adapter() })

describe(`EmptyRoomsScreen`, () => {
  let wrapper
  const mockNavigate = jest.fn()
  const mockProps = {
    navigation: {
      navigate: mockNavigate,
    },
  }

  const sampleRooms = [
    {
      classification_name: `Classroom`,
      roomname: `Bentham`,
      siteid: `b`,
      sitename: `Bentham House`,
    },
    {
      classification_name: `Lecture Theatre`,
      roomname: `Galactica`,
      siteid: `main`,
      sitename: `Main Building`,
    },
    {
      classification_name: `Classroom`,
      roomname: `Denning`,
      siteid: `main`,
      sitename: `Main Building`,
    },
  ]

  const mockGetEmptyRooms = jest.fn(() => Promise.resolve(sampleRooms))
  ApiManager.rooms.getEmptyRooms = mockGetEmptyRooms

  beforeEach(() => {
    wrapper = shallow(<EmptyRoomsScreen {...mockProps} />)
  })

  it(`renders the EmptyRoomsScreen`, () => {
    expect(wrapper).toMatchSnapshot()
  })

  it(`calls getEmptyRooms`, () => {
    jest.clearAllMocks()
    const instance = wrapper.instance()
    instance.componentDidMount()
    expect(mockGetEmptyRooms).toHaveBeenCalledTimes(1)
  })

  it(`shows message when no empty rooms found`, () => {
    wrapper.setState({ loadingEmptyRooms: false })
    expect(wrapper).toMatchSnapshot()
  })

  it(`supports filtering by building`, async () => {
    jest.clearAllMocks()
    const instance = wrapper.instance()
    await instance.fetchEmptyRooms()
    expect(instance.state.emptyRooms).toBe(sampleRooms)
    expect(instance.state.loadingEmptyRooms).toBe(false)
    expect(instance.state.sites).toHaveLength(2)

    const buildingPicker = wrapper.findWhere(
      (node) => node.prop(`testID`) === `buildingPicker`,
    )
    expect(buildingPicker.exists())
    buildingPicker.props().onValueChange(`main`)
    expect(instance.state.selectedSite).toBe(`main`)

    expect(wrapper).toMatchSnapshot()
  })
})
