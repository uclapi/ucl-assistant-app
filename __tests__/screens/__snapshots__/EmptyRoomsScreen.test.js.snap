// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyRoomsScreen renders the EmptyRoomsScreen 1`] = `
<Page
  bottomColour="#fefefe"
  contentContainerStyle={Object {}}
  keyboardAvoidingViewStyle={Object {}}
  mainTabPage={false}
  onRefresh={[Function]}
  refreshEnabled={false}
  refreshing={false}
  safeAreaViewStyle={Object {}}
  style={Object {}}
  topColour="#fefefe"
>
  <View
    style={
      Object {
        "paddingBottom": 20,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</Page>
`;

exports[`EmptyRoomsScreen shows message when no empty rooms found 1`] = `
<Page
  bottomColour="#fefefe"
  contentContainerStyle={Object {}}
  keyboardAvoidingViewStyle={Object {}}
  mainTabPage={false}
  onRefresh={[Function]}
  refreshEnabled={false}
  refreshing={false}
  safeAreaViewStyle={Object {}}
  style={Object {}}
  topColour="#fefefe"
>
  <View
    style={
      Object {
        "paddingBottom": 20,
      }
    }
  >
    <BodyText
      style={Object {}}
    >
      Filter by:
    </BodyText>
    <Picker
      mode="dialog"
      onValueChange={[Function]}
      selectedValue={null}
      testID="buildingPicker"
    >
      <PickerItem
        label="All Buildings"
        value={null}
      />
    </Picker>
    <BodyText
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      Rooms vacant for the next hour:
    </BodyText>
    <CentredText
      style={Object {}}
    >
      No empty rooms found :(
    </CentredText>
  </View>
</Page>
`;

exports[`EmptyRoomsScreen supports filtering by building 1`] = `
<Page
  bottomColour="#fefefe"
  contentContainerStyle={Object {}}
  keyboardAvoidingViewStyle={Object {}}
  mainTabPage={false}
  onRefresh={[Function]}
  refreshEnabled={false}
  refreshing={false}
  safeAreaViewStyle={Object {}}
  style={Object {}}
  topColour="#fefefe"
>
  <View
    style={
      Object {
        "paddingBottom": 20,
      }
    }
  >
    <BodyText
      style={Object {}}
    >
      Filter by:
    </BodyText>
    <Picker
      mode="dialog"
      onValueChange={[Function]}
      selectedValue="main"
      testID="buildingPicker"
    >
      <PickerItem
        label="All Buildings"
        value={null}
      />
      <PickerItem
        key="b"
        label="Bentham House"
        value="b"
      />
      <PickerItem
        key="main"
        label="Main Building"
        value="main"
      />
    </Picker>
    <BodyText
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      Rooms vacant for the next hour:
    </BodyText>
    <SearchResult
      bottomText="Lecture Theatre"
      buttonText="View"
      indicator={false}
      indicatorColor="#eaeaea"
      indicatorLoading={false}
      key="undefined-Galactica-main-Lecture Theatre"
      onPress={[Function]}
      style={Object {}}
      topText="Galactica"
      type="location"
    />
    <SearchResult
      bottomText="Classroom"
      buttonText="View"
      indicator={false}
      indicatorColor="#eaeaea"
      indicatorLoading={false}
      key="undefined-Denning-main-Classroom"
      onPress={[Function]}
      style={Object {}}
      topText="Denning"
      type="location"
    />
  </View>
</Page>
`;
